internal/test/compilecheck/compile_check_test.go

```diff
@@ -35,50 +88,130 @@ func TestCompileCheck(t *testing.T) {
 
 	rootDir := filepath.Dir(filepath.Dir(filepath.Dir(cwd)))
 
-	for _, version := range versions {
-		version := version // Capture range variable.
+	// Build the image and start one container we can reuse for all subtests.
+	req := testcontainers.ContainerRequest{
+		FromDockerfile: testcontainers.FromDockerfile{
+			Context:       rootDir,
+			Dockerfile:    "Dockerfile",
+			PrintBuildLog: true,
+		},
+		Entrypoint: []string{"tail", "-f", "/dev/null"},
+		WorkingDir: "/workspace",
+	}
+
+	genReq := testcontainers.GenericContainerRequest{ContainerRequest: req, Started: true}
+
+	container, err := testcontainers.GenericContainer(context.Background(), genReq)
+	require.NoError(t, err)
+
+	t.Cleanup(func() {
+		require.NoError(t, container.Terminate(context.Background()))
+	})
+
+	// Write main.go into the container.
+	exitCode, outputReader, err := container.Exec(context.Background(), []string{"sh", "-c", fmt.Sprintf("cat > /workspace/main.go << 'GOFILE'\n%s\nGOFILE", mainGo)})
+	require.NoError(t, err)
+
+	output, err := io.ReadAll(outputReader)
+	require.NoError(t, err)
+	require.Equal(t, 0, exitCode, "failed to write main.go: %s", output)
```

Comment:
[blocking] Use the io.Reader from [ContainerFile](https://pkg.go.dev/github.com/testcontainers/testcontainers-go#ContainerFile) to write main.go:

```go
func TestCompileCheck(t *testing.T) {
    // [...]

	req := testcontainers.ContainerRequest{
		FromDockerfile: testcontainers.FromDockerfile{
			Context:       filepath.Join(cwd, "docker"),
			Dockerfile:    "build.Dockerfile",
			PrintBuildLog: true,
		},
		Files: []testcontainers.ContainerFile{
			{
				Reader:            strings.NewReader(mainSrc),
				ContainerFilePath: "/workspace/main.go",
				FileMode:          0o644,
			},
		},
		Entrypoint: []string{"tail", "-f", "/dev/null"},
		WorkingDir: "/workspace",
	}

    // [...]
}
```